app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java

```diff
@@ -0,0 +1,281 @@
+package com.igalia.wolvic.browser.api.impl;
+
+import android.graphics.Bitmap;
+import android.graphics.Canvas;
+import android.graphics.Matrix;
+import android.graphics.Paint;
+import android.os.SystemClock;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.igalia.wolvic.browser.api.WImage;
+import com.igalia.wolvic.browser.api.WMediaSession;
+import com.igalia.wolvic.browser.api.WResult;
+
+import org.chromium.base.task.PostTask;
+import org.chromium.content_public.browser.MediaSession;
+import org.chromium.content_public.browser.MediaSessionObserver;
+import org.chromium.content_public.browser.UiThreadTaskTraits;
+import org.chromium.content_public.browser.WebContents;
+import org.chromium.components.browser_ui.media.MediaNotificationImageUtils;
+import org.chromium.components.browser_ui.media.MediaImageCallback;
+import org.chromium.components.browser_ui.media.MediaImageManager;
+import org.chromium.services.media_session.MediaImage;
+import org.chromium.services.media_session.MediaMetadata;
+import org.chromium.services.media_session.MediaPosition;
+
+import java.util.List;
+
+import org.chromium.media_session.mojom.MediaSessionAction;
+
+public class TabMediaSessionObserver extends MediaSessionObserver implements MediaImageCallback {
+    private static final int UPDATE_POSITION_TIME_MS = 500;
+
+    private @NonNull WebContents mWebContents;
+    private @NonNull SessionImpl mSession;
+    private WMediaSessionImpl mMediaSession = new WMediaSessionImpl();
+    private MediaImageManager mMediaImageManager;
+    private MediaMetadata mMetadata;
+    private WImage mMediaImage;
+    private MediaPosition mMediaPosition;
+    private boolean mIsActive = false;
+    private boolean mIsSuspended = false;
+    private boolean mRunUpdatingPositionTask = false;
+
+    public TabMediaSessionObserver(@NonNull WebContents webContents, @NonNull SessionImpl session) {
+        super(MediaSession.fromWebContents(webContents));
+
+        mWebContents = webContents;
```

Comment:
Should we keep the reference `mWebContents`? Because this private reference doesn't seem to use in this class except here.

Reply:
Done.